src/charm.py

```diff
@@ -0,0 +1,88 @@
+#!/usr/bin/env python3
+
+import logging
+from pathlib import Path
+from random import choices
+from string import ascii_uppercase, digits
+from typing import Union, Callable
+from ops.charm import CharmBase
+from ops.main import main
+from ops.model import ActiveStatus, MaintenanceStatus
+
+from oci_image import OCIImageResource, OCIImageResourceError
+
+log = logging.getLogger()
+
+
+def get_or_set(name: str, *, default: Union[str, Callable[[], str]]) -> str:
+    try:
+        path = Path(f"/run/{name}")
+        return path.read_text()
+    except FileNotFoundError:
+        value = default() if callable(default) else default
+        path.write_text(value)
+        return value
+
+
+def gen_pass() -> str:
+    return "".join(choices(ascii_uppercase + digits, k=30))
+
+
+class MinioCharm(CharmBase):
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.image = OCIImageResource(self, "oci-image")
+        self.framework.observe(self.on.install, self.set_pod_spec)
+        self.framework.observe(self.on.upgrade_charm, self.set_pod_spec)
+        self.framework.observe(self.on.config_changed, self.set_pod_spec)
+        self.framework.observe(self.on.minio_relation_joined, self.send_info)
+
+    def send_info(self, event):
+        secret_key = get_or_set("password", default=gen_pass)
+        event.relation.data[self.unit]["service"] = self.model.app.name
+        event.relation.data[self.unit]["port"] = self.model.config["port"]
```

Comment:
Tested this with the argo charms and this needs to be a string.  `str(self.model.config["port"])` will fix it.